core/src/main/scala/kafka/cluster/Replica.scala

```diff
@@ -98,31 +101,39 @@ class Replica(val brokerId: Int, val topicPartition: TopicPartition) extends Log
    * fetch request is always smaller than the leader's LEO, which can happen if small produce requests are received at
    * high frequency.
    */
-  def updateFetchState(
+  def maybeUpdateFetchState(
     followerFetchOffsetMetadata: LogOffsetMetadata,
     followerStartOffset: Long,
     followerFetchTimeMs: Long,
     leaderEndOffset: Long,
     brokerEpoch: Long
-  ): Unit = {
+  ): Boolean = {
+    var updateSuccess = true
     replicaState.updateAndGet { currentReplicaState =>
-      val lastCaughtUpTime = if (followerFetchOffsetMetadata.messageOffset >= leaderEndOffset) {
-        math.max(currentReplicaState.lastCaughtUpTimeMs, followerFetchTimeMs)
-      } else if (followerFetchOffsetMetadata.messageOffset >= currentReplicaState.lastFetchLeaderLogEndOffset) {
-        math.max(currentReplicaState.lastCaughtUpTimeMs, currentReplicaState.lastFetchTimeMs)
+      // Fence the update if it provides a stale broker epoch.
+      if (brokerEpoch != -1 && brokerEpoch < currentReplicaState.brokerEpoch.getOrElse(-1L)) {
+        updateSuccess = false
```

Comment:
It is not recommended to have side effect in the method updating the state of an atomic value. I wonder if we could throw an exception instead to really ensure that the update process is stopped.

Reply:
Do we need a KIP for the extra exception?

Reply:
Does it require a KIP to add a new exception?